File: flutter/storyly_flutter/android/src/main/kotlin/com/appsamurai/storyly/storyly_flutter/VerticalFeedData.kt
package com.appsamurai.storyly.storyly_flutter

import android.content.Context
import android.graphics.Color
import android.util.TypedValue
import com.appsamurai.storyly.StorylyLayoutDirection
import com.appsamurai.storyly.config.StorylyProductConfig
import com.appsamurai.storyly.config.StorylyShareConfig
import com.appsamurai.storyly.verticalfeed.StorylyVerticalFeedGroupOrder
import com.appsamurai.storyly.verticalfeed.StorylyVerticalFeedInit
import com.appsamurai.storyly.verticalfeed.VerticalFeedGroup
import com.appsamurai.storyly.verticalfeed.VerticalFeedItem
import com.appsamurai.storyly.verticalfeed.VerticalFeedItemButtonComponent
import com.appsamurai.storyly.verticalfeed.VerticalFeedItemCommentComponent
import com.appsamurai.storyly.verticalfeed.VerticalFeedItemComponent
import com.appsamurai.storyly.verticalfeed.VerticalFeedItemEmojiComponent
import com.appsamurai.storyly.verticalfeed.VerticalFeedItemPollComponent
import com.appsamurai.storyly.verticalfeed.VerticalFeedItemProductCardComponent
import com.appsamurai.storyly.verticalfeed.VerticalFeedItemProductCatalogComponent
import com.appsamurai.storyly.verticalfeed.VerticalFeedItemProductTagComponent
import com.appsamurai.storyly.verticalfeed.VerticalFeedItemPromoCodeComponent
import com.appsamurai.storyly.verticalfeed.VerticalFeedItemQuizComponent
import com.appsamurai.storyly.verticalfeed.VerticalFeedItemRatingComponent
import com.appsamurai.storyly.verticalfeed.VerticalFeedItemSwipeComponent
import com.appsamurai.storyly.verticalfeed.config.StorylyVerticalFeedConfig
import com.appsamurai.storyly.verticalfeed.config.bar.StorylyVerticalFeedBarStyling
import com.appsamurai.storyly.verticalfeed.config.customization.StorylyVerticalFeedCustomization
import com.appsamurai.storyly.verticalfeed.config.group.StorylyVerticalFeedGroupStyling
import java.util.Locale


internal class VerticalFeedInitMapper(
    private val context: Context,
) {

    fun getStorylyInit(
        json: HashMap<String, Any>
    ): StorylyVerticalFeedInit? {
        val storylyInitJson = json["storylyInit"] as? Map<String, *> ?: return null
        val storylyId = storylyInitJson["storylyId"] as? String ?: return null
        val storyGroupStylingJson = json["verticalFeedGroupStyling"] as? Map<String, *> ?: return null
        val storyBarStylingJson = json["verticalFeedBarStyling"] as? Map<String, *> ?: return null
        val storyStylingJson = json["verticalFeedCustomization"] as? Map<String, *> ?: return null
        val storyShareConfigJson = json["verticalFeedItemShareConfig"] as? Map<String, *> ?: return null
        val storyProductConfigJson = json["verticalFeedItemProductConfig"] as? Map<String, *> ?: return null

        var storylyConfigBuilder = StorylyVerticalFeedConfig.Builder()
        storylyConfigBuilder = stStorylyInit(json = storylyInitJson, configBuilder = storylyConfigBuilder)
        storylyConfigBuilder = stStorylyGroupStyling(context = context, json = storyGroupStylingJson, configBuilder = storylyConfigBuilder)
        storylyConfigBuilder = stStoryBarStyling(json = storyBarStylingJson, configBuilder = storylyConfigBuilder)
        storylyConfigBuilder = stStoryStyling(context = context, json = storyStylingJson, configBuilder = storylyConfigBuilder)
        storylyConfigBuilder = stShareConfig(json = storyShareConfigJson, configBuilder = storylyConfigBuilder)
        storylyConfigBuilder = stProductConfig(json = storyProductConfigJson, configBuilder = storylyConfigBuilder)


        return StorylyVerticalFeedInit(
            storylyId = storylyId,
            config = storylyConfigBuilder
                .build()
        )
    }

    private fun stStorylyInit(
        json: Map<String, *>,
        configBuilder: StorylyVerticalFeedConfig.Builder
    ): StorylyVerticalFeedConfig.Builder {
        return configBuilder
            .setLabels((json["storylySegments"] as? List<String>)?.toSet())
            .setCustomParameter(json["customParameter"] as? String)
            .setTestMode(json["storylyIsTestMode"] as? Boolean ?: false)
            .setUserData(json["userProperty"] as? Map<String, String> ?: emptyMap())
            .setLayoutDirection(getStorylyLayoutDirection(json["storylyLayoutDirection"] as? String))
            .setLocale(json["storylyLocale"] as? String)
    }

    private fun stStorylyGroupStyling(
        context: Context,
        json: Map<String, *>,
        configBuilder: StorylyVerticalFeedConfig.Builder,
    ): StorylyVerticalFeedConfig.Builder {
        var groupStylingBuilder = StorylyVerticalFeedGroupStyling.Builder()
        (json["iconBackgroundColor"] as? String)?.let { groupStylingBuilder = groupStylingBuilder.setIconBackgroundColor(
            Color.parseColor(it)) }
        (json["iconCornerRadius"] as? Int)?.let { groupStylingBuilder = groupStylingBuilder.setIconCornerRadius(it) }
        (json["iconHeight"] as? Int)?.let { groupStylingBuilder = groupStylingBuilder.setIconHeight(it) }
        (json["textColor"] as? String)?.let { groupStylingBuilder = groupStylingBuilder.setTextColor(
            Color.parseColor(it)) }
        groupStylingBuilder.setTypeface(getTypeface(context, json["titleFont"] as? String))
        (json["titleTextSize"] as? Int)?.let { groupStylingBuilder = groupStylingBuilder.setTitleTextSize(Pair(
            TypedValue.COMPLEX_UNIT_PX, it)) }
        groupStylingBuilder = groupStylingBuilder.setTitleVisibility(json["titleVisible"] as? Boolean ?: true)
        getVerticalFeedGroupOrder(json["groupOrder"] as? String)?.let { groupStylingBuilder = groupStylingBuilder.setGroupOrder(it) }
        (json["typeIndicatorVisible"] as? Boolean)?.let { groupStylingBuilder = groupStylingBuilder.setTypeIndicatorVisibility(it) }
        (json["minLikeCountToShowIcon"] as? Int)?.let { groupStylingBuilder = groupStylingBuilder.setMinLikeCountToShowIcon(it) }
        (json["minImpressionCountToShowIcon"] as? Int)?.let { groupStylingBuilder = groupStylingBuilder.setMinImpressionCountToShowIcon(it) }
        getDrawable(context, json["impressionIcon"] as? String)?.let { groupStylingBuilder = groupStylingBuilder.setImpressionIcon(it) }
        getDrawable(context, json["likeIcon"] as? String)?.let { groupStylingBuilder = groupStylingBuilder.setLikeIcon(it) }

        return configBuilder
            .setVerticalFeedGroupStyling(
                styling = groupStylingBuilder
                    .build()
            )
    }

    private fun stStoryBarStyling(
        json: Map<String, *>,
        configBuilder: StorylyVerticalFeedConfig.Builder
    ): StorylyVerticalFeedConfig.Builder {
        var barStyling = StorylyVerticalFeedBarStyling.Builder()
        (json["sections"] as? Int)?.let { barStyling = barStyling.setSection(it) }
        (json["horizontalEdgePadding"] as? Int)?.let { barStyling = barStyling.setHorizontalEdgePadding(it) }
        (json["verticalEdgePadding"] as? Int)?.let { barStyling = barStyling.setVerticalEdgePadding(it) }
        (json["horizontalPaddingBetweenItems"] as? Int)?.let { barStyling = barStyling.setHorizontalPaddingBetweenItems(it) }
        (json["verticalPaddingBetweenItems"] as? Int)?.let { barStyling = barStyling.setVerticalPaddingBetweenItems(it) }
        return configBuilder
            .setVerticalFeedBarStyling(
                barStyling.build()
            )
    }

    private fun stStoryStyling(
        context: Context,
        json: Map<String, *>,
        configBuilder: StorylyVerticalFeedConfig.Builder
    ): StorylyVerticalFeedConfig.Builder {
        var customization = StorylyVerticalFeedCustomization.Builder()
        getTypefaceOrNull(context, json["titleFont"] as? String)?.let { customization.setTitleTypeface(it) }
        getTypefaceOrNull(context, json["interactiveFont"] as? String)?.let { customization.setInteractiveTypeface(it) }
        (json["progressBarColor"] as? List<String>?)?.let { customization = customization.setProgressBarColor(it.map { hexColor -> Color.parseColor(hexColor) }) }
        (json["isTitleVisible"] as? Boolean)?.let { customization = customization.setTitleVisibility(it) }
        (json["isLikeButtonVisible"] as? Boolean)?.let { customization = customization.setLikeButtonVisibility(it) }
        (json["isCloseButtonVisible"] as? Boolean)?.let { customization = customization.setCloseButtonVisibility(it) }
        getDrawable(context, json["closeButtonIcon"] as? String)?.let { customization = customization.setCloseButtonIcon(it) }
        getDrawable(context, json["likeButtonIcon"] as? String)?.let { customization = customization.setLikeButtonIcon(it) }
        getDrawable(context, json["shareButtonIcon"] as? String)?.let { customization = customization.setShareButtonIcon(it) }

        return configBuilder
            .setVerticalFeedStyling(
                customization
                    .build()
            )
    }

    private fun stShareConfig(
        json: Map<String, *>,
        configBuilder: StorylyVerticalFeedConfig.Builder
    ): StorylyVerticalFeedConfig.Builder {
        var shareConfigBuilder = StorylyShareConfig.Builder()
        (json["storylyShareUrl"] as? String)?.let { shareConfigBuilder = shareConfigBuilder.setShareUrl(it) }
        (json["storylyFacebookAppID"] as? String)?.let { shareConfigBuilder = shareConfigBuilder.setFacebookAppID(it) }
        return configBuilder
            .setShareConfig(
                shareConfigBuilder
                    .build()
            )
    }

    private fun stProductConfig(
        json: Map<String, *>,
        configBuilder: StorylyVerticalFeedConfig.Builder
    ): StorylyVerticalFeedConfig.Builder {
        var productConfigBuilder = StorylyProductConfig.Builder()
        (json["isFallbackEnabled"] as? Boolean)?.let { productConfigBuilder = productConfigBuilder.setFallbackAvailability(it) }
        (json["isCartEnabled"] as? Boolean)?.let { productConfigBuilder = productConfigBuilder.setCartAvailability(it) }
        (json["productFeed"] as? HashMap<String, ArrayList<HashMap<String, Any?>>?>)?.let { productFeed ->
            val feed = productFeed.mapValues { entry ->
                entry.value?.let { productList ->
                    productList.map { createSTRProductItem(it) }
                } ?: emptyList()
            }
            productConfigBuilder = productConfigBuilder.setProductFeed(feed)
        }

        return configBuilder
            .setProductConfig(
                productConfigBuilder
                    .build()
            )
    }

    private fun getVerticalFeedGroupOrder(order: String?): StorylyVerticalFeedGroupOrder? {
        return when (order) {
            "static" -> StorylyVerticalFeedGroupOrder.Static
            "bySeenState" -> StorylyVerticalFeedGroupOrder.BySeenState
            else -> null
        }
    }

    private fun getStorylyLayoutDirection(layoutDirection: String?): StorylyLayoutDirection {
        return when (layoutDirection) {
            "ltr" -> StorylyLayoutDirection.LTR
            "rtl" -> StorylyLayoutDirection.RTL
            else -> StorylyLayoutDirection.LTR
        }
    }
}

internal fun createStoryGroupMap(storyGroup: VerticalFeedGroup): Map<String, *> {
    return mapOf(
        "id" to storyGroup.uniqueId,
        "title" to storyGroup.title,
        "index" to storyGroup.index,
        "seen" to storyGroup.seen,
        "iconUrl" to storyGroup.iconUrl,
        "feedList" to storyGroup.feedList.map { story -> createStoryMap(story) },
        "pinned" to storyGroup.pinned,
        "type" to storyGroup.type.ordinal,
        "name" to storyGroup.name,
        "nudge" to storyGroup.nudge
    )
}

internal fun createStoryMap(story: VerticalFeedItem): Map<String, *> {
    return mapOf(
        "id" to story.uniqueId,
        "title" to story.title,
        "name" to story.name,
        "index" to story.index,
        "seen" to story.seen,
        "currentTime" to story.currentTime,
        "previewUrl" to story.previewUrl,
        "actionUrl" to story.actionUrl,
        "verticalFeedItemComponentList" to story.verticalFeedItemComponentList?.map { component -> createStoryComponentMap(component) },
        "actionProducts" to (story.actionProducts ?: listOf()).map { product -> createSTRProductItemMap(product) }
    )
}

internal fun createStoryComponentMap(storyComponent: VerticalFeedItemComponent): Map<String, *> {
    when (storyComponent) {
        is VerticalFeedItemButtonComponent -> {
            return mapOf(
                "type" to storyComponent.type.name.lowercase(Locale.ENGLISH),
                "id" to storyComponent.id,
                "customPayload" to storyComponent.customPayload,
                "text" to storyComponent.text,
                "actionUrl" to storyComponent.actionUrl,
                "products" to (storyComponent.products ?: listOf()).map { product -> createSTRProductItemMap(product) }
            )
        }

        is VerticalFeedItemSwipeComponent -> {
            return mapOf(
                "type" to storyComponent.type.name.lowercase(Locale.ENGLISH),
                "id" to storyComponent.id,
                "customPayload" to storyComponent.customPayload,
                "text" to storyComponent.text,
                "actionUrl" to storyComponent.actionUrl,
                "products" to (storyComponent.products ?: listOf()).map { product -> createSTRProductItemMap(product) }
            )
        }

        is VerticalFeedItemProductTagComponent -> {
            return mapOf(
                "type" to storyComponent.type.name.lowercase(Locale.ENGLISH),
                "id" to storyComponent.id,
                "customPayload" to storyComponent.customPayload,
                "actionUrl" to storyComponent.actionUrl,
                "products" to (storyComponent.products ?: listOf()).map { product -> createSTRProductItemMap(product) }
            )
        }

        is VerticalFeedItemProductCardComponent -> {
            return mapOf(
                "type" to storyComponent.type.name.lowercase(Locale.ENGLISH),
                "id" to storyComponent.id,
                "customPayload" to storyComponent.customPayload,
                "text" to storyComponent.text,
                "actionUrl" to storyComponent.actionUrl,
                "products" to (storyComponent.products ?: listOf()).map { product -> createSTRProductItemMap(product) }
            )
        }

        is VerticalFeedItemProductCatalogComponent -> {
            return mapOf(
                "type" to storyComponent.type.name.lowercase(Locale.ENGLISH),
                "id" to storyComponent.id,
                "customPayload" to storyComponent.customPayload,
                "actionUrlList" to (storyComponent.actionUrlList ?: listOf()),
                "products" to (storyComponent.products ?: listOf()).map { product -> createSTRProductItemMap(product) }
            )
        }

        is VerticalFeedItemQuizComponent -> {
            return mapOf(
                "type" to storyComponent.type.name.lowercase(Locale.ENGLISH),
                "id" to storyComponent.id,
                "customPayload" to storyComponent.customPayload,
                "title" to storyComponent.title,
                "options" to storyComponent.options,
                "rightAnswerIndex" to storyComponent.rightAnswerIndex,
                "selectedOptionIndex" to storyComponent.selectedOptionIndex,
            )
        }

        is VerticalFeedItemPollComponent -> {
            return mapOf(
                "type" to storyComponent.type.name.lowercase(Locale.ENGLISH),
                "id" to storyComponent.id,
                "customPayload" to storyComponent.customPayload,
                "title" to storyComponent.title,
                "options" to storyComponent.options,
                "selectedOptionIndex" to storyComponent.selectedOptionIndex,
            )
        }

        is VerticalFeedItemEmojiComponent -> {
            return mapOf(
                "type" to storyComponent.type.name.lowercase(Locale.ENGLISH),
                "id" to storyComponent.id,
                "customPayload" to storyComponent.customPayload,
                "emojiCodes" to storyComponent.emojiCodes,
                "selectedEmojiIndex" to storyComponent.selectedEmojiIndex,
            )
        }

        is VerticalFeedItemRatingComponent -> {
            return mapOf(
                "type" to storyComponent.type.name.lowercase(Locale.ENGLISH),
                "id" to storyComponent.id,
                "customPayload" to storyComponent.customPayload,
                "emojiCode" to storyComponent.emojiCode,
                "rating" to storyComponent.rating,
            )
        }

        is VerticalFeedItemPromoCodeComponent -> {
            return mapOf(
                "type" to storyComponent.type.name.lowercase(Locale.ENGLISH),
                "id" to storyComponent.id,
                "customPayload" to storyComponent.customPayload,
                "text" to storyComponent.text
            )
        }

        is VerticalFeedItemCommentComponent -> {
            return mapOf(
                "type" to storyComponent.type.name.lowercase(Locale.ENGLISH),
                "id" to storyComponent.id,
                "customPayload" to storyComponent.customPayload,
                "text" to storyComponent.text
            )
        }

        else -> {
            return mapOf(
                "type" to storyComponent.type.name.lowercase(Locale.ENGLISH),
                "id" to storyComponent.id,
                "customPayload" to storyComponent.customPayload,
            )
        }
    }
}

File: flutter/storyly_flutter/ios/Classes/VerticalFeedData.swift
@_spi(InternalFramework) 
import Storyly
import UIKit

internal class VerticalFeedInitMapper {

    func getStorylyInit(json: [String: Any]) -> StorylyVerticalFeedInit? {
        guard let storylyInitJson = json["storylyInit"] as? [String: Any],
              let storylyId = storylyInitJson["storylyId"] as? String else {
            return nil
        }

        let storyGroupStylingJson = json["verticalFeedGroupStyling"] as? [String: Any]
        let storyBarStylingJson = json["verticalFeedBarStyling"] as? [String: Any]
        let storyStylingJson = json["verticalFeedCustomization"] as? [String: Any]
        let storyShareConfigJson = json["verticalFeedItemShareConfig"] as? [String: Any]
        let storyProductConfigJson = json["verticalFeedItemProductConfig"] as? [String: Any]

        var storylyConfigBuilder = StorylyVerticalFeedConfig.Builder()
        storylyConfigBuilder = stStorylyInit(json: storylyInitJson, configBuilder: &storylyConfigBuilder)
        storylyConfigBuilder = stStorylyGroupStyling(json: storyGroupStylingJson, configBuilder: storylyConfigBuilder)
        storylyConfigBuilder = stStoryBarStyling(json: storyBarStylingJson, configBuilder: storylyConfigBuilder)
        storylyConfigBuilder = stStoryStyling(json: storyStylingJson, configBuilder: storylyConfigBuilder)
        storylyConfigBuilder = stShareConfig(json: storyShareConfigJson, configBuilder: storylyConfigBuilder)
        storylyConfigBuilder = stProductConfig(json: storyProductConfigJson, configBuilder: storylyConfigBuilder)

        let storylyConfig = storylyConfigBuilder.build()
        storylyConfig.setFramework(framework: "flutter")

        return StorylyVerticalFeedInit(
            storylyId: storylyId,
            config: storylyConfig
        )
    }

    private func stStorylyInit(json: [String: Any], configBuilder: inout StorylyVerticalFeedConfig.Builder) -> StorylyVerticalFeedConfig.Builder {
        if let segmentsData = json["storylySegments"] as? [String] { configBuilder = configBuilder.setLabels(labels: Set(segmentsData)) }
        return configBuilder
            .setCustomParameter(parameter: json["customParameter"] as? String)
            .setTestMode(isTest: (json["storylyIsTestMode"] as? Bool) ?? false)
            .setUserData(data: json["userProperty"] as? [String: String] ?? [:])
            .setLayoutDirection(direction: getStorylyLayoutDirection(direction: json["storylyLayoutDirection"] as? String))
            .setLocale(locale: json["storylyLocale"] as? String)
    }

    private func stStorylyGroupStyling(json: [String: Any]?, configBuilder: StorylyVerticalFeedConfig.Builder) -> StorylyVerticalFeedConfig.Builder {
        guard let json = json else { return configBuilder }
        var groupStylingBuilder = StorylyVerticalFeedGroupStyling.Builder()
        if let iconBackgroundColor = json["iconBackgroundColor"] as? String {
            groupStylingBuilder = groupStylingBuilder.setIconBackgroundColor(color: UIColor(hexString: iconBackgroundColor))
        }
        if let iconCornerRadius = json["iconCornerRadius"] as? CGFloat {
            groupStylingBuilder = groupStylingBuilder.setIconCornerRadius(radius: iconCornerRadius)
        }
        if let iconHeight = json["iconHeight"] as? CGFloat {
            groupStylingBuilder = groupStylingBuilder.setIconHeight(height: iconHeight)
        }
        if let textColor = json["textColor"] as? String {
            groupStylingBuilder = groupStylingBuilder.setTextColor(UIColor(hexString: textColor))
        }
        if let titleFont = json["titleFont"] as? String {
            let fontSize = json["titleTextSize"] as? CGFloat ?? 12
            groupStylingBuilder = groupStylingBuilder.setFont(font: getCustomFont(typeface: titleFont as NSString, fontSize: fontSize))
        }
        if let titleVisible = json["titleVisible"] as? Bool {
            groupStylingBuilder = groupStylingBuilder.setTitleVisibility(isVisible: titleVisible)
        }
        
        if let groupOrder = json["groupOrder"] as? String,
           let order = getVerticalFeedGroupOrder(order: groupOrder) {
            groupStylingBuilder = groupStylingBuilder.setGroupOrder(order: order)
        }
        if let typeIndicatorVisible = json["typeIndicatorVisible"] as? Bool {
            groupStylingBuilder = groupStylingBuilder.setTypeIndicatorVisibility(typeIndicatorVisible)
        }
        if let minLikeCountToShowIcon = json["minLikeCountToShowIcon"] as? Int {
            groupStylingBuilder = groupStylingBuilder.setMinLikeCountToShowIcon(minLikeCountToShowIcon)
        }
        if let minImpressionCountToShowIcon = json["minImpressionCountToShowIcon"] as? Int {
            groupStylingBuilder = groupStylingBuilder.setMinImpressionCountToShowIcon(minImpressionCountToShowIcon)
        }
        if let impressionIcon = json["impressionIcon"] as? String,
           let icon = UIImage(named: impressionIcon) {
            groupStylingBuilder = groupStylingBuilder.setImpressionIcon(icon)
        }
        if let likeIcon = json["likeIcon"] as? String,
           let icon = UIImage(named: likeIcon) {
            groupStylingBuilder = groupStylingBuilder.setImpressionIcon(icon)
        }

        return configBuilder
            .setVerticalFeedGroupStyling(styling: groupStylingBuilder.build())
    }

    private func stStoryBarStyling(json: [String: Any]?, configBuilder: StorylyVerticalFeedConfig.Builder) -> StorylyVerticalFeedConfig.Builder {
        guard let json = json else { return configBuilder }
        var barStyling = StorylyVerticalFeedBarStyling.Builder()
        if let sections = json["sections"] as? Int {
            barStyling = barStyling.setSection(count: sections)
        }
        if let horizontalEdgePadding = json["horizontalEdgePadding"] as? CGFloat {
            barStyling = barStyling.setHorizontalEdgePadding(padding: horizontalEdgePadding)
        }
        if let verticalEdgePadding = json["verticalEdgePadding"] as? CGFloat {
            barStyling = barStyling.setVerticalEdgePadding(padding: verticalEdgePadding)
        }
        if let horizontalPaddingBetweenItems = json["horizontalPaddingBetweenItems"] as? CGFloat {
            barStyling = barStyling.setHorizontalPaddingBetweenItems(padding: horizontalPaddingBetweenItems)
        }
        if let verticalPaddingBetweenItems = json["verticalPaddingBetweenItems"] as? CGFloat {
            barStyling = barStyling.setVerticalPaddingBetweenItems(padding: verticalPaddingBetweenItems)
        }
        return configBuilder
            .setVerticalFeedBarStyling(styling: barStyling.build())
    }

    private func stStoryStyling(json: [String: Any]?, configBuilder: StorylyVerticalFeedConfig.Builder) -> StorylyVerticalFeedConfig.Builder {
        guard let json = json else { return configBuilder }
        var customization = StorylyVerticalFeedCustomization.Builder()
        if let titleFont = json["titleFont"] as? String,
           let font = getCustomFont(typeface: titleFont, fontSize: 14){
            customization = customization.setTitleFont(font: font)
        }
        if let interactiveFont = json["interactiveFont"] as? String,
           let font = getCustomFont(typeface: interactiveFont, fontSize: 14) {
            customization = customization.setInteractiveFont(font: font)
        }
        if let progressBarColor = json["progressBarColor"] as? [String] {
            customization = customization.setProgressBarColor(colors: progressBarColor.map { UIColor(hexString: $0) })
        }
        if let isTitleVisible = json["isTitleVisible"] as? Bool {
            customization = customization.setTitleVisibility(isVisible: isTitleVisible)
        }
        if let isLikeButtonVisible = json["isLikeButtonVisible"] as? Bool {
            customization = customization.setLikeButtonVisibility(isVisible: isLikeButtonVisible)
        }
        if let isCloseButtonVisible = json["isCloseButtonVisible"] as? Bool {
            customization = customization.setCloseButtonVisibility(isVisible: isCloseButtonVisible)
        }
        if let closeButtonIcon = json["closeButtonIcon"] as? String {
            customization = customization.setCloseButtonIcon(icon: UIImage(named: closeButtonIcon))
        }
        if let likeButtonIcon = json["likeButtonIcon"] as? String {
            customization = customization.setLikeButtonIcon(icon: UIImage(named: likeButtonIcon))
        }
        if let shareButtonIcon = json["shareButtonIcon"] as? String {
            customization = customization.setShareButtonIcon(icon: UIImage(named: shareButtonIcon))
        }

        return configBuilder
            .setVerticalFeedStyling(styling: customization.build())
    }

    private func stShareConfig(json: [String: Any]?, configBuilder: StorylyVerticalFeedConfig.Builder) -> StorylyVerticalFeedConfig.Builder {
        guard let json = json else { return configBuilder }
        var shareConfigBuilder = StorylyShareConfig.Builder()
        if let storylyShareUrl = json["storylyShareUrl"] as? String {
            shareConfigBuilder = shareConfigBuilder.setShareUrl(url: storylyShareUrl)
        }
        if let storylyFacebookAppID = json["storylyFacebookAppID"] as? String {
            shareConfigBuilder = shareConfigBuilder.setFacebookAppID(id: storylyFacebookAppID)
        }
        return configBuilder
            .setShareConfig(config: shareConfigBuilder.build())
    }

    private func stProductConfig(json: [String: Any]?, configBuilder: StorylyVerticalFeedConfig.Builder) -> StorylyVerticalFeedConfig.Builder {
        guard let json = json else { return configBuilder }
        var productConfigBuilder = StorylyProductConfig.Builder()
        if let isFallbackEnabled = json["isFallbackEnabled"] as? Bool {
            productConfigBuilder = productConfigBuilder.setFallbackAvailability(isEnabled: isFallbackEnabled)
        }
        if let isCartEnabled = json["isCartEnabled"] as? Bool {
            productConfigBuilder = productConfigBuilder.setCartEnabled(isEnabled: isCartEnabled)
        }
        if let productFeed = json["productFeed"] as? [String: [[String: Any]]] {
            let feed = productFeed.mapValues { productList in
                productList.map { createSTRProductItem(product: $0) }
            }
            productConfigBuilder = productConfigBuilder.setProductFeed(feed: feed)
        }

        return configBuilder
            .setProductConfig(config: productConfigBuilder.build())
    }

    private func getVerticalFeedGroupOrder(order: String?) -> StorylyVerticalFeedGroupOrder? {
        switch order {
        case "static":
            return .Static
        case "bySeenState":
            return .BySeenState
        default:
            return nil
        }
    }
    
    private func getStorylyLayoutDirection(direction: String?) -> StorylyLayoutDirection {
        switch direction {
        case "ltr": return .LTR
        case "rtl": return .RTL
        default: return .LTR
        }
    }
}

internal func createStoryGroupMap(storyGroup: VerticalFeedGroup?) -> [String: Any?]? {
    guard let storyGroup = storyGroup else { return nil }
    return [
        "id": storyGroup.uniqueId,
        "title": storyGroup.title,
        "index": storyGroup.index,
        "seen": storyGroup.seen,
        "iconUrl": storyGroup.iconUrl?.absoluteString,
        "feedList": storyGroup.feedList.map { story in createStoryMap(story: story) },
        "pinned": storyGroup.pinned,
        "type": storyGroup.type.rawValue,
        "name": storyGroup.name,
        "nudge": storyGroup.nudge
    ]
}

internal func createStoryMap(story: VerticalFeedItem?) -> [String: Any?]? {
    guard let story = story else { return nil }
    return [
        "id": story.uniqueId,
        "title": story.title,
        "name": story.name,
        "index": story.index,
        "seen": story.seen,
        "currentTime": story.currentTime,
        "previewUrl": story.previewUrl?.absoluteString,
        "actionUrl": story.actionUrl,
        "verticalFeedItemComponentList": story.verticalFeedItemComponentList?.map { component in createStoryComponentMap(storyComponent: component) },
        "actionProducts": (story.actionProducts ?? []).map { product in createSTRProductItemMap(product: product) }
    ]
}

internal func createStoryComponentMap(storyComponent: VerticalFeedItemComponent?) -> [String: Any?]? {
    guard let storyComponent = storyComponent else { return nil }
    switch storyComponent {
    case let buttonComponent as VerticalFeedButtonComponent:
        return [
            "type": "buttonaction",
            "id": buttonComponent.id,
            "customPayload": buttonComponent.customPayload,
            "text": buttonComponent.text,
            "actionUrl": buttonComponent.actionUrl,
            "products": (buttonComponent.products ?? []).compactMap { createSTRProductItemMap(product: $0) }
        ]
    case let swipeComponent as VerticalFeedSwipeComponent:
        return [
            "type": "swipeaction",
            "id": swipeComponent.id,
            "customPayload": swipeComponent.customPayload,
            "text": swipeComponent.text,
            "actionUrl": swipeComponent.actionUrl,
            "products": (swipeComponent.products ?? []).compactMap { createSTRProductItemMap(product: $0) }
        ]
    case let productTagComponent as VerticalFeedProductTagComponent:
        return [
            "type": "producttag",
            "id": productTagComponent.id,
            "customPayload": productTagComponent.customPayload,
            "actionUrl": productTagComponent.actionUrl,
            "products": (productTagComponent.products ?? []).compactMap { createSTRProductItemMap(product: $0) }
        ]
    case let productCardComponent as VerticalFeedProductCardComponent:
        return [
            "type": "productcard",
            "id": productCardComponent.id,
            "customPayload": productCardComponent.customPayload,
            "text": productCardComponent.text,
            "actionUrl": productCardComponent.actionUrl,
            "products": (productCardComponent.products ?? []).compactMap { createSTRProductItemMap(product: $0) }
        ]
    case let productCatalogComponent as VerticalFeedProductCatalogComponent:
        return [
            "type": "productcatalog",
            "id": productCatalogComponent.id,
            "customPayload": productCatalogComponent.customPayload,
            "actionUrlList": productCatalogComponent.actionUrlList ?? [],
            "products": (productCatalogComponent.products ?? []).compactMap { createSTRProductItemMap(product: $0) }
        ]
    case let quizComponent as VerticalFeedQuizComponent:
        return [
            "type": "quiz",
            "id": quizComponent.id,
            "customPayload": quizComponent.customPayload,
            "title": quizComponent.title,
            "options": quizComponent.options,
            "rightAnswerIndex": quizComponent.rightAnswerIndex?.intValue,
            "selectedOptionIndex": quizComponent.selectedOptionIndex,
        ]
    case let imageQuizComponent as VerticalFeedImageQuizComponent:
        return [
            "type": "imagequiz",
            "id": imageQuizComponent.id,
            "customPayload": imageQuizComponent.customPayload,
            "title": imageQuizComponent.title,
            "options": imageQuizComponent.options ?? [],
            "rightAnswerIndex": imageQuizComponent.rightAnswerIndex?.intValue,
            "selectedOptionIndex": imageQuizComponent.selectedOptionIndex,
        ]
    case let pollComponent as VerticalFeedPollComponent:
        return [
            "type": "poll",
            "id": pollComponent.id,
            "customPayload": pollComponent.customPayload,
            "title": pollComponent.title,
            "options": pollComponent.options,
            "selectedOptionIndex": pollComponent.selectedOptionIndex
        ]
    case let emojiComponent as VerticalFeedEmojiComponent:
        return [
            "type": "emoji",
            "id": emojiComponent.id,
            "customPayload": emojiComponent.customPayload,
            "emojiCodes": emojiComponent.emojiCodes,
            "selectedEmojiIndex": emojiComponent.selectedEmojiIndex
        ]
    case let ratingComponent as VerticalFeedRatingComponent:
        return [
            "type": "rating",
            "id": ratingComponent.id,
            "customPayload": ratingComponent.customPayload,
            "emojiCode": ratingComponent.emojiCode,
            "rating": ratingComponent.rating
        ]
    case let promoCodeComponent as VerticalFeedPromoCodeComponent:
        return [
            "type": "promocode",
            "id": promoCodeComponent.id,
            "customPayload": promoCodeComponent.customPayload,
            "text": promoCodeComponent.text
        ]
    case let commentComponent as VerticalFeedCommentComponent:
        return [
            "type": "comment",
            "id": commentComponent.id,
            "customPayload": commentComponent.customPayload,
            "text": commentComponent.text
        ]
    default:
        return [
            "type": VerticalFeedItemComponentTypeHelper.verticalFeedItemComponentName(componentType: storyComponent.type).lowercased(),
            "id": storyComponent.id,
            "customPayload": storyComponent.customPayload,
        ]
        
    }
}
